### PR TITLE
[Mips] In LowerShift*Parts, xor with bits-1 instead of -1.

### LINKED ISSUE #71142: [Mips] Incorrect code generated to perform i128 shifting by a truncated amount
### Description:

When compiling the following PoC with LLVM, an unexpected behavior happens, where the i128 representation of -1 will be shifted as if it were a <2xi64> vector with each element shifted by the given amount. In better-supported architectures and in older versions of LLVM, the shifting of i128 behaves correctly on Mips targets.

### Environment:

* **Compiler**: LLVM 18, commit b120fe8d3288c4dca1b5427ca34839ce8833f71c
* **Target Architecture**: mips64el-unknown-linux-gnuabi64
* **Optimization level**: the issue can be observed from O0 to O3

### PoC

```C
#include <stdio.h>
#include <stdint.h>

extern __int128 owo(__int128, int64_t);

int main() {
    __int128 a = 0xabcdef00abcdef11;
    a <<= 64;
    a |= 0xabcdef22abcdef33;
    __int128 b = owo(a, 48);
    unsigned long *p = (void *)&b;
    printf("%lx%lx\n", p[1], p[0]);
    return 0;
}
```

```llvm
target triple = "mips64el-unknown-linux-gnuabi64"

define i128 @owo(i128 %a, i64 %b) {
start:
  %tmp = sext i64 %b to i128
  %_1 = and i128 %tmp, 127
  %_0 = shl i128 %a, %_1
  ret i128 %_0
}

!1 = !{}
```

### Expected Behavior:

The resulted i128 should be `ef11abcdef22abcdef33000000000000`

### Observed Behavior:

The resulted i128 is `ef11000000000000ef33000000000000`

### Analysis:

The result looks like that when shifting by a number less than 64, although both i64 numbers get correctly shifted, the overflown bits of the lower 64 bits are not copied to the lower part of the higher 64 bits. Note that the and operation in the IR only applies to the number of bits to shift by, which is smaller than the limit of i8 in this case.

### PR DESCRIPTION
**Pull request by topperc**
If we start with an i128 shift, the initial shift amount would usually have zeros in bit 8 and above. xoring the shift amount with -1 will set those upper bits to 1. If DAGCombiner is able to prove those bits are now 1, then the shift that uses the xor will be replaced with undef. Which we don't want.

Reduce the xor constant to VT.bits-1 where VT is half the size of the larger shift type. This avoids toggling the upper bits. The hardware shift instruction only uses the lower bits of the shift amount. I assume the code used NOT because the hardware doesn't use the upper bits, but that isn't compatible with the LLVM poison semantics.

Fixes #71142.

I don't know who can review Mips anymore so I'm just adding mostly people who know about SelectionDAG and/or lowering shift parts.

### CODE REVIEW

**Comment by RKSimon** [comment]
FTR #71154 does something similar but this looks more complete

File: llvm/lib/Target/Mips/MipsISelLowering.cpp [hunk 1]
```diff
@@ -2593,12 +2593,13 @@ SDValue MipsTargetLowering::lowerShiftLeftParts(SDValue Op,
   SDValue Shamt = Op.getOperand(2);
   // if shamt < (VT.bits):
   //  lo = (shl lo, shamt)
-  //  hi = (or (shl hi, shamt) (srl (srl lo, 1), ~shamt))
+  //  hi = (or (shl hi, shamt) (srl (srl lo, 1), (VT.bits-1) ^ shamt))
```
Comment (RKSimon) [inline llvm/lib/Target/Mips/MipsISelLowering.cpp:2596]: consistent style: (xor (VT.bits-1), shamt)

File: llvm/lib/Target/Mips/MipsISelLowering.cpp [hunk 2]
```diff
@@ -2623,7 +2624,7 @@ SDValue MipsTargetLowering::lowerShiftRightParts(SDValue Op, SelectionDAG &DAG,
   MVT VT = Subtarget.isGP64bit() ? MVT::i64 : MVT::i32;
 
   // if shamt < (VT.bits):
-  //  lo = (or (shl (shl hi, 1), ~shamt) (srl lo, shamt))
+  //  lo = (or (shl (shl hi, 1), (VT.bits-1) ^ shamt) (srl lo, shamt))
```
Comment (RKSimon) [inline llvm/lib/Target/Mips/MipsISelLowering.cpp:2627]: consistent style: (xor (VT.bits-1), shamt)

**Review by RKSimon — APPROVED** [review]
LGTM (with a couple of minors)

**Comment by topperc** [comment]
There's another bug in this code that neither patch addresses.

I believe this violates the boolean rules. ISD::SELECT should only receive 0/1 as a condition. There needs to be a setcc here.
```
  SDValue Cond = DAG.getNode(ISD::AND, DL, MVT::i32, Shamt,                      
                             DAG.getConstant(VT.getSizeInBits(), DL, MVT::i32)); 
  Lo = DAG.getNode(ISD::SELECT, DL, VT, Cond,                                    
                   DAG.getConstant(0, DL, VT), ShiftLeftLo); 
```

**Comment by yingopq** [comment]
> There's another bug in this code that neither patch addresses.
> 
> I believe this violates the boolean rules. ISD::SELECT should only receive 0/1 as a condition. There needs to be a setcc here.
> 
> ```
>   SDValue Cond = DAG.getNode(ISD::AND, DL, MVT::i32, Shamt,                      
>                              DAG.getConstant(VT.getSizeInBits(), DL, MVT::i32)); 
>   Lo = DAG.getNode(ISD::SELECT, DL, VT, Cond,                                    
>                    DAG.getConstant(0, DL, VT), ShiftLeftLo); 
> ```

There has a question about shift number 129, now we deal with it as shift number 1 and the result of `shl i128 %a, %b` was `ffffffffffffffff fffffffffffffffe`.
If we add a setcc like the following diff, the above result was `fffffffffffffffe 0000000000000000` like RISCV `RISCVTargetLowering::lowerShiftLeftParts`.  Whether we're going to change the behavior of the shift number 129?
```diff
+//  SDValue Cond = DAG.getNode(ISD::AND, DL, MVT::i32, Shamt,
+//                             DAG.getConstant(VT.getSizeInBits(), DL, MVT::i32));
+
+  SDValue MinusXLen = DAG.getConstant(-(int)VT.getSizeInBits(), DL, MVT::i32);
+  SDValue ShamtMinusXLen = DAG.getNode(ISD::ADD, DL, MVT::i32, Shamt, MinusXLen);
+  SDValue Zero = DAG.getConstant(0, DL, MVT::i32);
+  SDValue CC = DAG.getSetCC(DL, MVT::i32, ShamtMinusXLen, Zero, ISD::SETLT);
+  Lo = DAG.getNode(ISD::SELECT, DL, VT, CC, ShiftLeftLo, DAG.getConstant(0, DL, VT));
+  Hi = DAG.getNode(ISD::SELECT, DL, VT, CC, Or, ShiftLeftLo);

```

**Pull request merged**